### PR TITLE
US - 0.5.0 - add/remove followers to musical entity

### PR DESCRIPTION
**Pull request by med-amine-bj**

### CODE REVIEW

File: Usiqee/Service/ServiceArtist.swift [hunk 1]
```diff
@@ -50,13 +50,40 @@ class ServiceArtist {
...
+        guard let user = Auth.auth().currentUser,
+              let follower = user.toFollower,
+              let artistId = artist.id else {
             return
         }
         
-        FFirestoreReference.userFollowedArtists(userId: user.uid).addDocument(data: artist.relatedData) { error in
+        let dispatchGroup = DispatchGroup()
+        var didFollowArtist: Bool = true
+        var didAddFollower: Bool = true
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:61]: Can we replace this two bool by hasError bool

File: Usiqee/Service/ServiceArtist.swift [hunk 2]
```diff
@@ -50,13 +50,40 @@ class ServiceArtist {
     }
     
     static func follow(artist: Artist, completion: @escaping (Error?) -> Void)  {
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:52]: can we have empty completion handler

Comment (guaire94) [reply Usiqee/Service/ServiceArtist.swift:52]: I can see error isn't use in the VC

File: Usiqee/Service/ServiceArtist.swift [hunk 3]
```diff
@@ -50,13 +50,40 @@ class ServiceArtist {
...
+        var didAddFollower: Bool = true
+        
+        dispatchGroup.enter()
+        FFirestoreReference.userFollowedArtists(userId: user.uid).addDocument(data: artist.toRelated) { error in
+            if let error = error {
+                didFollowArtist = false
+                completion(error)
+                dispatchGroup.leave()
+                return
+            }
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:70]: can we replace  that by hasError = true

File: Usiqee/Service/ServiceArtist.swift [hunk 4]
```diff
@@ -50,13 +50,40 @@ class ServiceArtist {
...
+        }
+        
+        dispatchGroup.enter()
+        FFirestoreReference.artistFollowers(artistId: artistId).document(user.uid).setData(follower.toRelated, completion: { error in
             if let error = error {
+                didAddFollower = false
                 completion(error)
+                dispatchGroup.leave()
+                return
+            }
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:81]: can we replace that by hasError = true

File: Usiqee/Service/ServiceArtist.swift [hunk 5]
```diff
@@ -50,13 +50,40 @@ class ServiceArtist {
...
+            dispatchGroup.leave()
+        })
+        
+        dispatchGroup.notify(queue: .main) {
+            guard didAddFollower, didFollowArtist else {
                 return
             }
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:88]:
```suggestion
            guard !hasError else {
                completion()
                return
            }
```

File: Usiqee/Service/ServiceArtist.swift [hunk 6]
```diff
@@ -73,9 +100,34 @@ class ServiceArtist {
...
+                didRemoveFollower = false
+                completion(error)
+                dispatchGroup.leave()
+                return
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.notify(queue: .main) {
+            guard didUnfollowArtist, didRemoveFollower else {
```
Comment (guaire94) [inline Usiqee/Service/ServiceArtist.swift:130]: same behaviour here

File: Usiqee/Service/ServiceBand.swift [hunk 7]
```diff
@@ -50,14 +50,40 @@ class ServiceBand {
     }
     
     static func follow(band: Band, completion: @escaping (Error?) -> Void)  {
```
Comment (guaire94) [inline Usiqee/Service/ServiceBand.swift:52]: same behaviour here

**Review by guaire94 — COMMENTED** [review]
What about unfollow band ?

File: Usiqee/Service/ServiceArtist.swift (hunk 2 above)
Comment (med-amine-bj) [reply Usiqee/Service/ServiceArtist.swift:52]: it's used in `ArtistDetailsVC`
````
ServiceArtist.follow(artist: artist, completion: { [weak self] error in
    self?.didFinishFollowing(error)
})
````

**Comment by med-amine-bj** [comment]
> What about unfollow band ?

je vois pas de quoi tu parle 🙄

**Pull request merged**